File: ruby/testcode/authzfailure_019.rb
require_relative 'shared'

# vuln-code-snippet start ruby_authz_bulk_delete
def bulk_delete(req)
  ids = req.param('ids').split(',')
  # No ownership check on bulk operation
  BenchmarkResponse.ok("deleted: #{ids.join(', ')}") # vuln-code-snippet vuln-line ruby_authz_bulk_delete
end
# vuln-code-snippet end ruby_authz_bulk_delete

File: rust/testcode/authzfailure_019.rs
//! CWE-285: Role check occurs after the sensitive action has already been performed

fn perform_action(data: &str) -> String {
    format!("action_performed_with_{}", data)
}

fn is_admin() -> bool {
    false
}

fn log_unauthorized() {
    let _ = "unauthorized_access_logged";
}

// vuln-code-snippet start testcodeAuthzfailure019
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let data = req.param("data");
    let result = perform_action(&data); // vuln-code-snippet target-line testcodeAuthzfailure019
    if !is_admin() {
        log_unauthorized();
    }
    super::shared::BenchmarkResponse::ok(&result)
}
// vuln-code-snippet end testcodeAuthzfailure019
